lib/HLSL/DxilOutputColorBecomesConstant.cpp

```diff
@@ -150,18 +133,128 @@ bool DxilOutputColorBecomesConstant::runOnModule(Module &M)
+    hasIntOutputs = true;
+  });
+
+    if (!hasFloatOutputs && !hasIntOutputs)
+  {
+    return false;
+  }
+
+  // Otherwise, we assume the shader outputs only one or the other (because the 0th RTV can't have a mixed type)
+  assert(hasFloatOutputs || hasIntOutputs);
+
+  std::array<llvm::Value *, 4> ReplacementColors;
+
+  switch (Mode)
+  {
+    case FromLiteralConstant: {
+      if (hasFloatOutputs) {
+        ReplacementColors[0] = HlslOP->GetFloatConst(Red);
+        ReplacementColors[1] = HlslOP->GetFloatConst(Green);
+        ReplacementColors[2] = HlslOP->GetFloatConst(Blue);
+        ReplacementColors[3] = HlslOP->GetFloatConst(Alpha);
+      }
+      if (hasIntOutputs) {
+        ReplacementColors[0] = HlslOP->GetI32Const(static_cast<int>(Red));
+        ReplacementColors[1] = HlslOP->GetI32Const(static_cast<int>(Green));
+        ReplacementColors[2] = HlslOP->GetI32Const(static_cast<int>(Blue));
+        ReplacementColors[3] = HlslOP->GetI32Const(static_cast<int>(Alpha));
+      }
+    }
+    break;
+    case FromConstantBuffer: {
+
+      // Setup a constant buffer with a single float4 in it:
+      SmallVector<llvm::Type*, 4> Elements { Type::getFloatTy(Ctx), Type::getFloatTy(Ctx) , Type::getFloatTy(Ctx) , Type::getFloatTy(Ctx) };
+      llvm::StructType *CBStructTy = llvm::StructType::create(Elements, "PIX_ConstantColorCB_Type");
+      std::unique_ptr<DxilCBuffer> pCBuf = llvm::make_unique<DxilCBuffer>();
+      pCBuf->SetGlobalName("PIX_ConstantColorCBName");
+      pCBuf->SetGlobalSymbol(UndefValue::get(CBStructTy));
+      pCBuf->SetID(0);
+      pCBuf->SetSpaceID((unsigned int)-2); // This is the reserved-for-tools register space
+      pCBuf->SetLowerBound(0);
+      pCBuf->SetRangeSize(1);
+      pCBuf->SetSize(4);
+
+      Instruction * entryPointInstruction = &*(DM.GetEntryFunction()->begin()->begin());
+      IRBuilder<> Builder(entryPointInstruction);
+
+      // Create handle for the newly-added constant buffer (which is achieved via a function call)
+      auto ConstantBufferName = "PIX_Constant_Color_CB_Handle";
+      Function *createHandle = HlslOP->GetOpFunc(DXIL::OpCode::CreateHandle, Type::getVoidTy(Ctx));
+      Constant *CreateHandleOpcodeArg = HlslOP->GetU32Const((unsigned)DXIL::OpCode::CreateHandle);
+      Constant *CBVArg = HlslOP->GetI8Const(static_cast<std::underlying_type<DxilResourceBase::Class>::type>(DXIL::ResourceClass::CBuffer)); 
+      Constant *MetaDataArg = HlslOP->GetU32Const(0); // position of the metadata record in the corresponding metadata list
+      Constant *IndexArg = HlslOP->GetU32Const(0); // 
+      Constant *FalseArg = HlslOP->GetI1Const(0); // non-uniform resource index: false
+      CallInst *callCreateHandle = Builder.CreateCall(createHandle, { CreateHandleOpcodeArg, CBVArg, MetaDataArg, IndexArg, FalseArg }, ConstantBufferName);
+
+      pCBuf->SetHandle(callCreateHandle);
+
+      ID = DM.AddCBuffer(std::move(pCBuf));
```

Comment:
Move ID = DM.AddCBuffer(std::mov(pCBuf)); up so 
MetaDataArg could use the ID.
Constant *MetaDataArg = HlslOP->GetU32Const(ID); //